src/java/org/apache/cassandra/db/compaction/ArenaSelector.java

```diff
@@ -0,0 +1,237 @@
+    {
+        return shardFor(key, shardBoundaries);
+    }
+
+    public static int shardFor(DecoratedKey key, List<PartitionPosition> shardBoundaries)
+    {
+        int pos = Collections.binarySearch(shardBoundaries, key);
+        assert pos < 0; // boundaries are .minkeybound and .maxkeybound so they should never be equal to a DecoratedKey
+        return -pos - 1;
+    }
+
+    public static int shardsSpanned(SSTableReader rdr, List<PartitionPosition> shardBoundaries)
+    {
+        if (shardBoundaries.size() <= 1)
+            return 1;
+        int startIdx = shardFor(rdr.getFirst(), shardBoundaries);
+        DecoratedKey last = rdr.getLast();
+        if (last.compareTo(shardBoundaries.get(startIdx)) < 0)
+            return 1;   // quick path, end boundary is in the same shard
+        return shardFor(last, shardBoundaries) - startIdx + 1;
+    }
+
+    public long shardAdjustedSize(SSTableReader rdr)
+    {
+        return shardAdjustedSize(rdr, shardBoundaries);
+    }
+
+    public static long shardAdjustedSize(SSTableReader rdr, List<PartitionPosition> shardBoundaries)
+    {
+        // This may need to duplicate the above to avoid the division in the happy path
+        return rdr.onDiskLength() / shardsSpanned(rdr, shardBoundaries);
+    }
+
+    public static Set<SSTableReader> sstablesFor(int boundaryIndex, List<PartitionPosition> shardBoundaries, Set<SSTableReader > sstables)
+    {
+        assert boundaryIndex < shardBoundaries.size();
+        return sstables.stream()
+                       .filter(sstable -> shardFor(sstable.getFirst(), shardBoundaries) <= boundaryIndex && shardFor(sstable.getLast(), shardBoundaries) >= boundaryIndex)
+                       .collect(Collectors.toSet());
+    }
+
+    public int compareByShardAdjustedSize(SSTableReader a, SSTableReader b)
+    {
+        return Long.compare(shardAdjustedSize(a), shardAdjustedSize(b));
+    }
+
+    /**
+     * An equivalence class is a function that compares two sstables and returns 0 when they fall in the same class.
+     * For example, the repair status or disk index may define equivalence classes. See the concrete equivalence classes below.
+     */
+    private interface EquivClassSplitter extends Comparator<SSTableReader> {
+
+        @Override
+        int compare(SSTableReader a, SSTableReader b);
+
+        /** Return a name that describes the equivalence class */
+        String name(SSTableReader ssTableReader);
+    }
+
+    /**
+     * Split sstables by their repair state: repaired, unrepaired, pending repair with a specific UUID (one group per pending repair).
+     */
+    private static final class RepairEquivClassSplitter implements EquivClassSplitter
+    {
+        public static final EquivClassSplitter INSTANCE = new RepairEquivClassSplitter();
+
+        @Override
+        public int compare(SSTableReader a, SSTableReader b)
+        {
+            // This is the same as name(apply(a)).compareTo(name(apply(b)))
```

Comment:
Removed `apply` from the comment in `compare` and the special condition for transient from `name` in 2e61bf9df3a1c473fd21f983cd4fe25b2002e1b7. I removed the special condition for transient because it is redundant - all SSTables within a single repair session are either transient or not and there is already a special arena for every session.